### DISCUSSION TITLE
How to share extension context across modules?

### DISCUSSION THREAD

**Discussion by zeluspudding** [microsoft/vscode-discussions#730] [Extension Development QnA] [answered]
I'm developing a VS Code extension using Javascript instead of Typescript. That said, I want to be able to use my extension's context across various modules without having to pass it as an explicit argument to each function in my modules (which is what I'm currently doing... the problem is I have a seeming Russian Doll situation of passing the context to a function that calls another function that calls another function which is the only function that needs the context).

My current approach relies on returning the context from the main `activate()` function [as advised here](https://stackoverflow.com/questions/62036883/how-to-get-extensioncontext-in-vscode-extension-unit-test). But this seems like **bad practice** since I'd be exposing my context to other extensions which could then access my sensitive information stored via secrets API, as [emphasized here](https://stackoverflow.com/a/70669712/3491991).

So how can we just load the context on a module or intra-function basis? Any code samples I find that seem to accomplish this all use Typescript (like [this sample](https://dev.to/kompotkot/how-to-use-secretstorage-in-your-vscode-extensions-2hco), or [this sample](https://github.com/microsoft/vscode-pull-request-github/blob/611d819b535b98c0066d0ef7e42e32bcb2a2b24d/src/github/credentials.ts#L40) ) which I can't get to work in my project.

**Comment by isidorn** [comment]
Use the [SecretStorage API](https://code.visualstudio.com/api/references/vscode-api#SecretStorage) that VS Code provides on the [ExtensionContext](https://code.visualstudio.com/api/references/vscode-api#ExtensionContext). This API is cross-platform and will work on all platforms that VS Code supports. It is also maintained by the VS Code team and will continue to be maintained and has been a part of the VS Code API for years at this point.

If that does not work can you share more details what exactly is the issue

**Reply by zeluspudding** [reply]
For whatever reason that just wouldn't work (not sure if this was particular to my mocha tests or if it was an issue across the board). The secrets API was always undefined as if it wouldn't initialize (even though I had via `activate(context)` ) . In any case, the globals approach as described below resolves my issue. Marking closed but feel free to reopen if you want me to dig up more details.

**Comment by r3m0t** [comment, marked as answer]
Have you tried setting the `ExtensionContext` as a module global and then importing that module from other modules? `import shared from "shared.js";` then read or write `shared.extensionContext`

You could use typescript without typechecking.
